### PR TITLE
[X86] Add i8 CTPOP lowering using i32 MUL

### LINKED ISSUE #79823: popcount 8-bit optimization opportunity (without HW instructions)
When compiling for x86-64 processor without POPCNT instruction, Clang generates inline code for `__builtin_popcount`. For counting bits of 8-bit only input, I have a smaller version than what Clang could generate. Are LLVM/Clang developers interested in incorporating this?

```c
#include <stdint.h>

unsigned int popCount8(uint8_t x) {
    return (unsigned int)__builtin_popcount(x);
}
unsigned int popCount8_b(uint8_t x) {
    // The algorithm is specific to 8-bit input, and avoids using 64-bit register for code size.
    uint32_t n = (uint32_t)(x * 0x08040201U);
    n = (uint32_t)(((n >> 3) & 0x11111111U) * 0x11111111U) >> 28;
    return n;
}
```

Clang's assembly output (`-Os -mno-popcnt`)
```
popCount8:
 mov    eax,edi
 shr    al,1
 and    al,0x55
 sub    dil,al
 mov    eax,edi
 and    al,0x33
 shr    dil,0x2
 and    dil,0x33
 add    dil,al
 mov    eax,edi
 shr    al,0x4
 add    al,dil
 and    al,0xf
 movzx  eax,al
 ret
popCount8_b:
 imul   eax,edi,0x8040201
 shr    eax,0x3
 and    eax,0x11111111
 imul   eax,eax,0x11111111
 shr    eax,0x1c
 ret
```

### PR DESCRIPTION
**Pull request by RKSimon**
This is first basic proposal in #79823 - we can investigate improving support for other widths if we can find further use cases.

### CODE REVIEW

File: llvm/test/CodeGen/X86/popcnt.ll [hunk 1]
```diff
@@ -59,16 +46,13 @@ define i8 @cnt8(i8 %x) nounwind readnone {
...
-; X64-NDD-NEXT:    shrb $4, %al, %cl
-; X64-NDD-NEXT:    addb %cl, %al
-; X64-NDD-NEXT:    andb $15, %al
+; X64-NDD-NEXT:    movzbl %dil, %eax
+; X64-NDD-NEXT:    imull $134480385, %eax, %eax # imm = 0x8040201
+; X64-NDD-NEXT:    shrl $3, %eax
+; X64-NDD-NEXT:    andl $286331153, %eax # imm = 0x11111111
+; X64-NDD-NEXT:    imull $286331153, %eax, %eax # imm = 0x11111111
+; X64-NDD-NEXT:    shrl $28, %eax
+; X64-NDD-NEXT:    # kill: def $al killed $al killed $eax
```
Comment (phoebewang) [inline llvm/test/CodeGen/X86/popcnt.ll:55]: Any chance to merge the three together?

File: llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp [hunk 2]
```diff
@@ -8639,6 +8639,24 @@ SDValue TargetLowering::expandCTPOP(SDNode *Node, SelectionDAG &DAG) const {
...
+    Op = DAG.getNode(ISD::SRL, dl, MVT::i32, Op,
+                     DAG.getShiftAmountConstant(3, MVT::i32, dl));
+    Op = DAG.getNode(ISD::AND, dl, MVT::i32, Op, Mask11);
+    Op = DAG.getNode(ISD::MUL, dl, MVT::i32, Op, Mask11);
+    Op = DAG.getNode(ISD::SRL, dl, MVT::i32, Op,
+                     DAG.getShiftAmountConstant(28, MVT::i32, dl));
+    return DAG.getZExtOrTrunc(Op, dl, MVT::i8);
+  }
+
   // This is the "best" algorithm from
```
Comment (phoebewang) [inline llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:8660]: So it is not the "best" algorithm now :)

**Review by phoebewang — APPROVED** [review]
LGTM.

File: llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp [hunk 3]
```diff
@@ -8639,6 +8639,24 @@ SDValue TargetLowering::expandCTPOP(SDNode *Node, SelectionDAG &DAG) const {
   if (VT.isVector() && !canExpandVectorCTPOP(*this, VT))
     return SDValue();
 
+  // i8 CTPOP - with efficient i32 MUL, then attempt multiply-mask-multiply.
+  if (VT == MVT::i8 && shouldAllowMultiplyInBitCounts(MVT::i8, MVT::i32) &&
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:8643]: Do we really need yet another extremely specific target hook? I would expect the isOperationLegal checks be the base implementation if so

Comment (RKSimon) [reply llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:8643]: So far only x86 (and maybe arm) have benefited from this

Comment (nikic) [reply llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp]: Would checking the immediate cost be able to distinguish targets where this is profitable?

**Comment by RKSimon** [comment]
Limited this to X86

**Review by phoebewang — APPROVED** [review]
LGTM.

File: llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp (hunk 2 above)
Comment (Explorer09) [reply llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp]: The algorithm is not the "best" in all cases. Although it does work for the common cases of counting the 32-bit and 64-bit integers. :)

**Pull request merged**